Adds a dbt version compatibility tool (#333)
snowplow/documentation: docs/modeling-your-data/modeling-your-data-with-dbt/index.md

</TabItem>
</Tabs>

## dbt Version Compatibility Checker

We understand that you may be using other dbt packages that require you to use a specific version of dbt, or a specific version of the `dbt_utils` package, or you simply may not wish to upgrade your existing setup. For that reason we provide a way to check what is the latest version of our packages that you can install for a given setup. Simply enter your dbt and (optional) `dbt_utils` version below.

```mdx-code-block
import { dbtVersions } from '@site/src/dbtVersions';
import { satisfies, gt, valid } from "semver";
import { useState } from 'react';
import TextField from '@mui/material/TextField';

> 
> 
> 

Reviewer: ```suggestion
```

export function VersionChecker() {
    const [activeIndex, setActiveIndex] = useState(0);
    const [dbtVer, setDbtVer] = useState('')
    const [utilsVer, setutilsVer] = useState('')
    return (
        <>

>         <TextField

Reviewer: indentation is a bit messed up here and below

            id="input-dbt-version"
            label="dbt Version"
            onChange={() => setDbtVer(event.target.value)}
            />

>             &nbsp; &nbsp;

Author: Just looks ugly without them and the text boxes touching

            <TextField
            id="input-utils-version"
            label="dbt-utils Version"
            onChange={() => setutilsVer(event.target.value)}
            />
        <br/>
        <br/>
        <Getsupportedpackages dbtVer={dbtVer} utilsVer={utilsVer}/>
        </>
    );
}

> export const Getsupportedpackages = ({children, dbtVer, utilsVer}) => {

Reviewer: ```suggestion
export const GetSupportedPackages = ({children, dbtVer, utilsVer}) => {
```
We are not savages, are we? :)

    let packageVersions = []
    if (! valid(dbtVer)) {

>         return(<>
>         <i>Please enter a valid dbt version.</i>
>         </>)

Reviewer: ```suggestion
        return(<em>Please enter a valid dbt version.</em>)
```

    }
    if (! valid(utilsVer) && utilsVer != '') {
        return(<>
        <i>Please enter a valid dbt-utils version.</i>

>         </>)

Reviewer: Same as above

    }
    if (utilsVer === '') {
        for(const [pkg, version] of Object.entries(dbtVersions)) {
            let maxVer = '0.0.0'
            let actualMaxVer = '0.0.0'
            for(const [ver, details] of Object.entries(version)) {

>                 {/* Check if the version is in the range the package supports, and check if this version is newer than our previous highest */}

Reviewer: ```suggestion
                // Check if the version is in the range the package supports, and check if this version is newer than our previous highest
```

>                 if (satisfies(dbtVer, details['dbtversion'], {'includePrerelease':true}) && gt(ver, maxVer) ) {

Reviewer: ```suggestion
                if (satisfies(dbtVer, details.dbtversion, {'includePrerelease':true}) && gt(ver, maxVer) ) {
```

                    maxVer = ver
                }
                if (gt(ver, actualMaxVer)) {
                    actualMaxVer = ver
                }
            }

>             if (maxVer == '0.0.0') {

Reviewer: ```suggestion
            if (maxVer === '0.0.0') {
```

>                 packageVersions.push({'package': pkg, 'maxVer': 'none', 'actualMaxVer' : actualMaxVer})

Reviewer: ```suggestion
                packageVersions.push({'package': pkg, 'maxVer': null, actualMaxVer})
```

            } else {

>                 packageVersions.push({'package': pkg, 'maxVer': maxVer, 'actualMaxVer' : actualMaxVer})

Reviewer: ```suggestion
                packageVersions.push({'package': pkg, maxVer, actualMaxVer})
```

Reviewer: Maybe rename `pkg` for the same effect?

            }
        }
        return (
        <>

>         <p>For <b>dbt</b> version <code>{dbtVer}</code> the latest version of each of our packages you can install are:</p>

Reviewer: Indent

>         <p>For <b>dbt</b> version <code>{dbtVer}</code> the latest version of each of our packages you can install are:</p>

Reviewer: ```suggestion
        <p>For <strong>dbt</strong> version <code>{dbtVer}</code> the latest version of each of our packages you can install are:</p>
```

>         <p>For <b>dbt</b> version <code>{dbtVer}</code> the latest version of each of our packages you can install are:</p>

Reviewer: ```suggestion
        <p>For <strong>dbt</strong> version <code>{dbtVer}</code> the latest versions of our packages you can install are:</p>
```

>         {packageVersions.map(x => <li>{x['package']}: <code>{x['maxVer'] == 'none' ? 'No matching version supported' : x['maxVer'] }</code> &nbsp; &nbsp; <i>(latest version: <code>{x['actualMaxVer']}</code>)</i></li>)}

Reviewer: I think you forgot a surrounding `<ul>`

>         {packageVersions.map(x => <li>{x['package']}: <code>{x['maxVer'] == 'none' ? 'No matching version supported' : x['maxVer'] }</code> &nbsp; &nbsp; <i>(latest version: <code>{x['actualMaxVer']}</code>)</i></li>)}

Reviewer: ```suggestion
        {packageVersions.map(x => <li>{x.package}: <code>{x.maxVer || 'No matching version supported'}</code> &nbsp;<em>(latest version: <code>{x.actualMaxVer}</code>)</em></li>)}
```

        </>
        )
    } else {

>         for(const [pkg, version] of Object.entries(dbtVersions)) {

Reviewer: 1. Same comments as above
2. This seems to be 99% the same, can we avoid duplication?

Author: I got rid of most of it, I can't work out how to conditionally add the string in the first line of the return so there's still some duplication there...

            let maxVer = '0.0.0'
            let actualMaxVer = '0.0.0'
            for(const [ver, details] of Object.entries(version)) {
                {/* Check if the dbt version is in the range the package supports, AND if the dbt-utils version is within the range (if exists) and check if this version is newer than our previous highest */}
                let dbt_utils_required_version = details['packages']['dbt-labs/dbt_utils'] ?? utilsVer

>                 console.log(`Package ${pkg} required dbt utils ${dbt_utils_required_version}`)

Reviewer: Remove logging

</TabItem>
</Tabs>

## dbt Version Compatibility Checker

You may be using other dbt packages that require you to use a specific version of dbt, or a specific version of the `dbt_utils` package. Or you simply may not wish to upgrade your existing setup.

Here’s a way to check the latest version of our packages you can install for a given setup. Simply enter your dbt version and (optionally) `dbt_utils` version below.

```mdx-code-block
import { dbtVersions } from '@site/src/dbtVersions';
import { satisfies, gt, valid } from "semver";
import { useState } from 'react';
import TextField from '@mui/material/TextField';

export function VersionChecker() {
  const [activeIndex, setActiveIndex] = useState(0);
  const [dbtVer, setDbtVer] = useState('')
  const [utilsVer, setutilsVer] = useState('')
  return (
    <>

>     <TextField

Reviewer: Indentation still off, this should be indented (since it’s inside `<>`)

      id="input-dbt-version"
      label="dbt Version"
      onChange={() => setDbtVer(event.target.value)}
    />
    &nbsp; &nbsp;
    <TextField
      id="input-utils-version"
      label="dbt-utils Version"
      onChange={() => setutilsVer(event.target.value)}
    />
    <br/>
    <br/>
    <GetSupportedPackages dbtVer={dbtVer} utilsVer={utilsVer}/>
    </>
  );
}

export const GetSupportedPackages = ({children, dbtVer, utilsVer}) => {
  let packageVersions = []
  if (! valid(dbtVer)) {
    return(<em>Please enter a valid dbt version.</em>)
  }
  if (! valid(utilsVer) && utilsVer !== '') {
    return(<em>Please enter a valid dbt version.</em>)
  }
  for(const [pkg, version] of Object.entries(dbtVersions)) {
    let maxVer = '0.0.0'
    let actualMaxVer = '0.0.0'
    for(const [ver, details] of Object.entries(version)) {
      // Check if the version is in the range the package supports, ( AND if the dbt-utils version is within the range (if exists) ) and check if this version is newer than our previous highest

>       let dbt_utils_required_version = details['packages']['dbt-labs/dbt_utils'] ?? utilsVer

Reviewer: ```suggestion
      let dbt_utils_required_version = details.packages['dbt-labs/dbt_utils'] ?? utilsVer
```

>       if (utilsVer === '' && (satisfies(dbtVer, details.dbtversion, {'includePrerelease':true}) && gt(ver, maxVer)) ) {

Reviewer: ```suggestion
      if (utilsVer === '' && (satisfies(dbtVer, details.dbtversion, {includePrerelease:true}) && gt(ver, maxVer)) ) {
```

        maxVer = ver

>       } else if (utilsVer !== '' && (satisfies(dbtVer, details['dbtversion'], {'includePrerelease':true}) && satisfies(utilsVer, dbt_utils_required_version, {'includePrerelease':true}) && gt(ver, maxVer) )){

Reviewer: ```suggestion
      } else if (utilsVer !== '' && (satisfies(dbtVer, details.dbtversion, {includePrerelease:true}) && satisfies(utilsVer, dbt_utils_required_version, {includePrerelease:true}) && gt(ver, maxVer) )){
```

>       } else if (utilsVer !== '' && (satisfies(dbtVer, details['dbtversion'], {'includePrerelease':true}) && satisfies(utilsVer, dbt_utils_required_version, {'includePrerelease':true}) && gt(ver, maxVer) )){

Reviewer: Looks overcomplicated, `if (A && (!B || C))` replaces both conditions

Reviewer: ```
const satisfiesDbt = satisfies(dbtVer, details.dbtversion, {'includePrerelease':true})
const satisfiesDbtUtils = (utilsVer === '') || satisfies(utilsVer, dbt_utils_required_version, {includePrerelease:true})
if (satisfiesDbt && satisfiesDbtUtils && gt(ver, maxVer)) { ... }
```

        maxVer = ver
      }
      if (gt(ver, actualMaxVer)) {
        actualMaxVer = ver
      }
    }
    if (maxVer === '0.0.0') {
      packageVersions.push({pkg, 'maxVer': null, actualMaxVer})
    } else {
      packageVersions.push({pkg, maxVer, actualMaxVer})
    }
  }
  if (utilsVer !== '') {
    return (
      <>

>       <p>For <strong>dbt</strong> version <code>{dbtVer}</code> and <strong>dbt_utils</strong> version <code>{utilsVer}</code> the latest version of each of our packages you can install are:</p>

Reviewer: ```suggestion
      <p>For <strong>dbt</strong> version <code>{dbtVer}</code> and {utilsVer !== '' && (<><strong>dbt_utils</strong> version <code>{utilsVer}</code></>)} the latest version of each of our packages you can install are:</p>
```
Something like this
